crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -2321,6 +2333,20 @@ where
                         }
 
                         // Ex) List[...]
+                        if self
+                            .settings
+                            .rules
+                            .enabled(Rule::MissingFutureAnnotationsImport)
+                            && (self.settings.target_version < PythonVersion::Py39
+                                && (self.settings.target_version >= PythonVersion::Py37
+                                    && !self.ctx.future_annotations()
+                                    && self.ctx.in_annotation()))
+                            && analyze::typing::is_pep585_builtin(expr, &self.ctx)
```

Comment:
This condition is a lot to read, I'd like to clean this up... But the idea is: mirror the logic below for `Rule::NonPEP585Annotation`, but flip the `!self.ctx.future_annotations()`. So we're asking: would we trigger here if `__future__` annotations were enabled?